src/runtime_src/core/tools/common/JSONConfigurable.cpp

```diff
@@ -6,37 +6,62 @@
 #include "tools/common/JSONConfigurable.h"
 #include "core/common/error.h"
 
-// Find children corresponding to target(s) and return them in a ptree.
-boost::property_tree::ptree
-JSONConfigurable::parse_configuration(
-  const std::vector<std::string>& targets,
+#include <iostream>
+#include <set>
+
+const std::map<std::string, std::string>
+JSONConfigurable::device_type_map = {
+  {"aie", "AIE"},
+  {"alveo", "Alveo"},
+  {"common", "common"}
+};
+
+// Parse configuration for children corresponding to current device categories before parsing for target(s).
+std::map<std::string, boost::property_tree::ptree>
+JSONConfigurable::parse_configuration_tree(
   const boost::property_tree::ptree& configuration)
 {
-  boost::property_tree::ptree output;
-  try {
-    for (const std::string& target: targets) {
-      for (const auto& configTree : configuration) {
-        for (const auto& config : configTree.second.get_child("contents")) {
-          const auto& pt = config.second;
-          if (boost::iequals(target, pt.get<std::string>("name")))
-            output.push_back({ configTree.second.get<std::string>("name"), pt });
+  std::map<std::string, boost::property_tree::ptree> target_mappings;
+  for (const auto& device_config_tree : configuration) { // Thanks ptree JSON parsing -_- We LOVE nested for loops!
+    for (const auto& device_config : device_config_tree.second) {
+      for (const auto& command_config_tree : device_config.second) {
+        for (const auto& command_config : command_config_tree.second) {
+          auto it = target_mappings.find(command_config.first);
+          if (it == target_mappings.end()) {
+            boost::property_tree::ptree target_tree;
+            target_tree.push_back({device_config.first, command_config.second});
+            target_mappings.emplace(command_config.first, target_tree);
+          } else
+            it->second.push_back({device_config.first, command_config.second});
         }
       }
     }
-  } catch (const std::exception& e) {
-    throw std::runtime_error(e.what());
   }
-  return output;
+
+  return target_mappings;
 }
 
-// Parse configuration for children corresponding to current device categories before parsing for target(s).
-boost::property_tree::ptree
-JSONConfigurable::parse_configuration_tree(
-  const std::vector<std::string>& targets,
-  const boost::property_tree::ptree& configuration)
+std::set<std::shared_ptr<JSONConfigurable>>
+JSONConfigurable::extract_common_options(std::map<std::string, std::vector<std::shared_ptr<JSONConfigurable>>> device_options)
```

Comment:
std::map by value, really?
Is shared ptr really what you want to use here.  What is a JSONConfigurable ? Is ownership really shared here?

Reply:
Converted as much as I could to `std::vector<T*>`